circuit-macros/src/circuit_type/singleprover_circuit_types.rs

```diff
@@ -0,0 +1,240 @@
+pub(crate) fn build_circuit_types(base_type: &ItemStruct) -> TokenStream2 {
+    let mut res_stream = TokenStream2::default();
+
+    // Build an implementation of `CircuitBaseType` for the
+    let circuit_base_type_stream = build_circuit_base_type_impl(base_type);
+    res_stream.extend(circuit_base_type_stream);
+
+    // Build a variable type
+    let var_type_stream = build_var_type(base_type);
+    res_stream.extend(var_type_stream);
+
+    // Build a commitment types
+    let comm_type_stream = build_commitment_type(base_type);
+    res_stream.extend(comm_type_stream);
+
+    res_stream
+}
+
+/// Build an `impl CircuitBaseType` block for the base type
+fn build_circuit_base_type_impl(base_type: &ItemStruct) -> TokenStream2 {
+    let base_name = base_type.ident.clone();
+    let trait_ident = new_ident(CIRCUIT_BASE_TYPE_TRAIT_NAME);
+
+    let var_type_associated = new_ident(VAR_TYPE_ASSOCIATED_NAME);
+    let var_type_generics = build_var_type_generics();
+    let var_type_name =
+        with_var_type_generics(ident_with_suffix(&base_name.to_string(), VAR_TYPE_SUFFIX));
+    let comm_type_associated = new_ident(COMM_TYPE_ASSOCIATED_NAME);
+    let comm_type_name = ident_with_suffix(&base_name.to_string(), COMM_TYPE_SUFFIX);
+
+    // Build the body of the `commitment_randomness` method
+    let commitment_randomness_impl = build_commitment_randomness_method(base_type);
+    let impl_block: ItemImpl = parse_quote! {
+        impl #trait_ident for #base_name {
+            type #var_type_associated #var_type_generics = #var_type_name;
+            type #comm_type_associated = #comm_type_name;
+
+            #commitment_randomness_impl
+        }
+    };
+    impl_block.to_token_stream()
+}
+
+/// Build an implementation of the `commitment_randomness` method that calls out to each
+/// field's implementation
+pub(crate) fn build_commitment_randomness_method(base_type: &ItemStruct) -> TokenStream2 {
+    // Build the body of the `commitment_randomness` method
+    let commitment_randomness_ident = new_ident(COMMITMENT_RANDOMNESS_METHOD_NAME);
+    let mut field_stmts: Vec<Stmt> = Vec::new();
+    for field in base_type.fields.iter() {
+        let field_ident = field.ident.clone();
+        field_stmts.push(parse_quote! {
+            res.extend(self.#field_ident.#commitment_randomness_ident(r));
+        });
+    }
+
+    let fn_def: ItemFn = parse_quote! {
+        fn #commitment_randomness_ident <R: RngCore + CryptoRng>(&self, r: &mut R) -> Vec<Scalar> {
+            let mut res = Vec::new();
+            #(#field_stmts)*
+
+            res
+        }
+    };
+    fn_def.to_token_stream()
+}
+
+/// Build a variable type; the type of the base allocated in a constraint system
+pub(crate) fn build_var_type(base_type: &ItemStruct) -> TokenStream2 {
+    let base_name = base_type.ident.clone();
+    let var_name = ident_with_suffix(&base_name.to_string(), VAR_TYPE_SUFFIX);
+    let derive_clone: Attribute = parse_quote!(#[derive(Clone)]);
```

Comment:
We can put it on the base type, it doesn't make sense to have default on a variable type though

Reply:
i'm thinking about making tests easier to write down the line, would be nice to have a default witness / statement / variable / commitment / etc. for basic unit tests

Reply:
also selfishly am thinking about being able to easily make dummy data so that i can build the constraint system for the export tool, but I feel like adding `Default` derives is 1) a small lift, and 2) doesn't really violate abstraction boundaries

Reply:
woah, good point - thanks for explaining

Reply:
oh and fwiw, i guess it's pretty easy to make dummy data by doing `BaseType::default().commit()`, since the `Commit{Public, Verifier, Witness, etc.}` traits will be auto-derived in your next PR